File: deploy/check-env-drift.py
#!/usr/bin/env python3
"""check-env-drift.py — does an island's LIVE .env match the repo's encrypted copy?

THE ONE GAP THIS CLOSES, and it is the whole scope. `preflight-compose-drift.sh`
already refuses a deploy when the box's `docker-compose.yml` or `deploy/` differ
from the tag being pulled. It cannot see `.env`, BY CONSTRUCTION: `.env` holds
per-box secrets and there is nothing public to compare it against. So half the
config surface — the half that held `APNS_VOIP_TOPIC` on 2026-09-11 while compose
failed to forward it — has never been checkable at all.

This makes it checkable. That is all it does.

WHAT IT DELIBERATELY DOES NOT DO, and why the list is this long. A design that
SHIPPED `.env` to the box went four rounds of `/design-temper` and never reached
SOUND (`docs/design/16-TEMPER.md`). Every finding across those rounds — the pin
that could not leave `.env`, the backup landing in a pruned generation, a stopped
tenant read as a missing one, an empty pin file silently meaning `edge`, a
half-applied discriminator that compared two names of the same live symlink — was
about WRITING, or about the generation directory that writing required. None of
them survive a read-only tool.

So: it never writes to the box. No generation, no symlink, no restore, no
baseline, no adoption, no cutover, and nothing new on the box to sync or prune.
It refuses and shows, exactly as ISL-0003's drift guard does, for ISL-0003's
stated reason: a box may legitimately carry local state, and silently clobbering
it is how #2301 became standing instead of caught.

NAMES ONLY, NEVER VALUES. `deploy/secrets/MANIFEST.txt` already established this
convention and its reasoning; this tool matches it. It reports WHICH KEYS differ,
never what they differ to. `--show-values` exists for an operator who has already
decided to look, and it is not the default.

NO TEMP FILE, NOTHING TO SHRED. The plaintext is read over a pipe from `sops -d`
and held only in this process. It never reaches a file, never reaches argv, and
never reaches the shell environment — which were the three exfil surfaces the v1
temper named (`/proc/<pid>/environ` is readable, `set -x` echoes, children
inherit, and `trap` does not run on SIGKILL). Removing the file removes the
shredding problem rather than solving it.

Usage:
    deploy/check-env-drift.py <island> [<island> ...] [--show-values]
    deploy/check-env-drift.py --all

Exit codes — three outcomes, and the third fails closed:
    0  MATCH         every key present on both sides with equal values
    1  DIFFERS       a real difference; the key names are on stdout
    2  COULD NOT RUN ssh failed, sops failed, no key, wrong box, parse failed

A check that could not run is not a check that passed.
"""

from __future__ import annotations

import argparse
import subprocess
import sys
from pathlib import Path

REPO = Path(__file__).resolve().parent.parent

# The two live islands. Kept here rather than in a conf file the tool would have
# to parse: there is no `deploy/islands/` in this repo, and inventing one to hold
# two rows would be the kind of machinery four temper rounds spent their time
# deleting. A third island adds a row.
ISLANDS = {
    "enspyr": {"ssh": "nick-mel", "path": "~/apps/aiko-chat-gateway"},
    "imagineering": {"ssh": "imagineering", "path": "~/apps/aiko-chat-gateway"},
}

EXIT_MATCH, EXIT_DIFFERS, EXIT_CANNOT_RUN = 0, 1, 2


class CannotRun(Exception):
    """Anything that stops the comparison happening. Never confused with a clean run."""


def parse_dotenv(text: str, where: str) -> dict[str, str]:
    """Parse with the SAME library the gateway parses with.

    `python-dotenv` is already a runtime dependency, and using it means "these
    two agree" means what it means at boot. Hand-rolling a parser here would
    reintroduce the risk it exists to measure — and `APNS_PRIVATE_KEY` is a PEM
    with REAL newlines inside quotes, which is exactly why SOPS's own dotenv
    parser rejects these files and they are stored as binary blobs.

    KNOWN LIMIT, stated rather than discovered: Compose's interpolation parser is
    not this parser. A value those two disagree about would read as MATCH here.
    That is a narrower gap than the one this closes, and it is not silently
    assumed away.
    """
    try:
        from dotenv import dotenv_values
    except ImportError as exc:  # pragma: no cover - environment problem, not logic
        raise CannotRun(
            "python-dotenv is not importable. Run from the repo venv: "
            f".venv/bin/python {Path(__file__).name} …"
        ) from exc

    import io

    parsed = dotenv_values(stream=io.StringIO(text))
    out = {k: v for k, v in parsed.items() if v is not None}
    if not out:
        raise CannotRun(f"parsed zero keys from {where} — refusing to call that a match")
    return out


def decrypted_repo_env(island: str) -> str:
    """`sops -d` over a pipe. Plaintext never touches a file, argv or the environment."""
    path = REPO / "deploy" / "secrets" / f"{island}.env.sops"
    if not path.is_file():
        raise CannotRun(f"no encrypted config at {path}")
    try:
        proc = subprocess.run(
            ["sops", "-d", str(path)],
            capture_output=True,
            text=True,
            timeout=60,
            check=False,
        )
    except FileNotFoundError as exc:
        raise CannotRun("sops is not installed on this machine (control-side only)") from exc
    except subprocess.TimeoutExpired as exc:
        raise CannotRun("sops timed out — is it waiting on a key?") from exc
    if proc.returncode != 0:
        raise CannotRun(f"sops could not decrypt {path.name}: {proc.stderr.strip()[:400]}")
    return proc.stdout


def live_env(island: str) -> str:
    """Read the box's `.env`. READ ONLY — this is the only thing touching the box."""
    cfg = ISLANDS[island]
    try:
        proc = subprocess.run(
            ["ssh", "-o", "BatchMode=yes", cfg["ssh"], f"cat {cfg['path']}/.env"],
            capture_output=True,
            text=True,
            timeout=60,
            check=False,
        )
    except subprocess.TimeoutExpired as exc:
        raise CannotRun(f"ssh to {cfg['ssh']} timed out") from exc
    if proc.returncode != 0:
        raise CannotRun(f"could not read {cfg['path']}/.env on {cfg['ssh']}: {proc.stderr.strip()[:400]}")
    return proc.stdout


def assert_same_island(island: str, repo: dict[str, str], live: dict[str, str]) -> None:
    """Confirm the box is the island we decrypted for, BEFORE reporting anything.

    Both boxes are shared, multi-tenant hosts, and a wrong-box comparison would
    report a wall of spurious differences that reads exactly like catastrophic
    drift. Cheap, and it is the read-only sibling of the tenant preflight the
    temper rounds demanded of the shipping design.
    """
    for key in ("ISLAND_ID", "ISLAND_DISPLAY_NAME"):
        a, b = repo.get(key), live.get(key)
        if a and b and a != b:
            raise CannotRun(
                f"WRONG BOX: {key} differs between the repo's {island} config and "
                f"{ISLANDS[island]['ssh']}. Refusing to compare."
            )


def compare(repo: dict[str, str], live: dict[str, str]) -> tuple[list[str], list[str], list[str]]:
    only_repo = sorted(set(repo) - set(live))
    only_live = sorted(set(live) - set(repo))
    changed = sorted(k for k in set(repo) & set(live) if repo[k] != live[k])
    return only_repo, only_live, changed


def report(island: str, only_repo, only_live, changed, repo, live, show_values: bool) -> None:
    ssh = ISLANDS[island]["ssh"]
    print(f"\n=== {island} ({ssh}) ===")
    if not (only_repo or only_live or changed):
        print("  MATCH — the box's .env agrees with the repo's encrypted copy.")
        return
    print("  DIFFERS")
    if only_repo:
        print(f"\n  in the REPO but NOT on the box ({len(only_repo)}):")
        for k in only_repo:
            print(f"    - {k}")
    if only_live:
        print(f"\n  on the BOX but NOT in the repo ({len(only_live)}):")
        for k in only_live:
            print(f"    + {k}")
    if changed:
        print(f"\n  present in both, VALUES DIFFER ({len(changed)}):")
        for k in changed:
            if show_values:
                print(f"    ~ {k}\n        repo: {repo[k]!r}\n        box:  {live[k]!r}")
            else:
                print(f"    ~ {k}")
    if changed and not show_values:
        print("\n  (key names only — pass --show-values to print the values themselves)")


def check(island: str, show_values: bool) -> int:
    repo = parse_dotenv(decrypted_repo_env(island), f"repo copy of {island}")
    live = parse_dotenv(live_env(island), f"live .env on {ISLANDS[island]['ssh']}")
    assert_same_island(island, repo, live)
    only_repo, only_live, changed = compare(repo, live)
    report(island, only_repo, only_live, changed, repo, live, show_values)
    return EXIT_DIFFERS if (only_repo or only_live or changed) else EXIT_MATCH


def main() -> int:
    ap = argparse.ArgumentParser(
        description="Does an island's live .env match the repo's encrypted copy? Read-only.",
        epilog="Exit: 0 match, 1 differs, 2 could not run (fails closed).",
    )
    ap.add_argument("islands", nargs="*", choices=[*ISLANDS, []], help="island name(s)")
    ap.add_argument("--all", action="store_true", help="check every known island")
    ap.add_argument(
        "--show-values",
        action="store_true",
        help="print the differing values, not just their key names (default: names only)",
    )
    args = ap.parse_args()

    targets = list(ISLANDS) if args.all else args.islands
    if not targets:
        ap.error("name at least one island, or pass --all")

    worst = EXIT_MATCH
    for island in targets:
        try:
            worst = max(worst, check(island, args.show_values))
        except CannotRun as exc:
            print(f"\n=== {island} ===\n  COULD NOT RUN — {exc}", file=sys.stderr)
            worst = EXIT_CANNOT_RUN
    return worst


if __name__ == "__main__":
    sys.exit(main())

File: deploy/dart/bin/check_env_drift.dart
#!/usr/bin/env dart

/// check_env_drift — does an island's LIVE `.env` match the repo's encrypted copy?
///
/// THE ONE GAP THIS CLOSES, and it is the whole scope.
/// `preflight-compose-drift.sh` already refuses a deploy when the box's
/// `docker-compose.yml` or `deploy/` differ from the tag being pulled. It cannot
/// see `.env` BY CONSTRUCTION: `.env` holds per-box secrets and there is nothing
/// public to compare it against. So half the config surface — the half that held
/// `APNS_VOIP_TOPIC` on 2026-09-11 while compose failed to forward it — has
/// never been checkable at all. This makes it checkable. That is all it does.
///
/// WHAT IT DELIBERATELY DOES NOT DO. A design that SHIPPED `.env` to the box
/// went four rounds of `/design-temper` and never reached SOUND
/// (`docs/design/16-TEMPER.md`). Every finding across those rounds — the pin
/// that could not leave `.env`, the backup landing in a pruned generation, a
/// stopped tenant read as a missing one, an empty pin file silently meaning
/// `edge`, a half-applied discriminator comparing two names of the same live
/// symlink — was about WRITING, or about the generation directory writing
/// required. None of them survive a read-only tool. So: it never writes to the
/// box. No generation, no symlink, no restore, no baseline, no adoption, no
/// cutover, nothing new on the box to sync or prune.
///
/// NAMES ONLY, NEVER VALUES, matching `deploy/secrets/MANIFEST.txt`'s
/// convention. `--show-values` exists for an operator who has already decided to
/// look, and it is not the default.
///
/// WHY DART (#4824 Stage 1, first item; Nick 2026-09-26 "dart rewrite first").
/// This supersedes the unmerged `deploy/check-env-drift.py` of PR#190 — the
/// issue names it as the free one to redo, since redoing an unmerged file is not
/// a migration. The honest account of what the language buys is in
/// `lib/drift.dart`: the three outcomes are a sealed type, so the exit-code
/// mapping cannot drift from the set of things that can happen, and a switch
/// that forgets COULD-NOT-RUN does not compile. It buys nothing in
/// `lib/dotenv.dart`, which is a parser, and that file says so.
///
/// Usage:
///   `dart run deploy/dart/bin/check_env_drift.dart <island> [...] [--show-values]`
///   dart run deploy/dart/bin/check_env_drift.dart --all
///
/// Exit codes — three outcomes, and the third fails closed:
///   0  MATCH          every key present on both sides with equal values
///   1  DIFFERS        a real difference; the key names are on stdout
///   2  COULD NOT RUN  ssh failed, sops failed, no key, wrong box, parse failed
///
/// A check that could not run is not a check that passed.
library;

import 'dart:io';

import 'package:island_deploy/dotenv.dart';
import 'package:island_deploy/drift.dart';
import 'package:island_deploy/islands.dart';

Future<void> main(List<String> argv) async {
  final showValues = argv.contains('--show-values');
  final all = argv.contains('--all');
  final named = argv.where((a) => !a.startsWith('--')).toList();

  if (!all && named.isEmpty) {
    stderr.writeln(_usage);
    exit(2);
  }
  final unknown = named.where((n) => !islands.containsKey(n)).toList();
  if (unknown.isNotEmpty) {
    stderr.writeln('unknown island(s): ${unknown.join(", ")}\n$_usage');
    exit(2);
  }

  final targets = all ? islands.keys.toList() : named;
  final repoRoot = _repoRoot();

  var worst = 0;
  for (final name in targets) {
    final result = await check(repoRoot, islands[name]!);
    _report(result, showValues);
    final code = exitCodeFor(result.outcome);
    if (code > worst) worst = code;
  }
  exit(worst);
}

/// An outcome plus, only for a [Differs], the two sides' values.
///
/// The values ride HERE rather than on [Differs] so the outcome object an
/// operator might log or pass around can never carry a secret. An earlier draft
/// smuggled them through a module-level variable written by `check` and read by
/// the reporter — correct only because the loop happened to call them in that
/// order, which is a property of today's caller rather than of the code.
typedef CheckResult = ({
  CheckOutcome outcome,
  Map<String, (String, String)> values,
});

const _noValues = <String, (String, String)>{};

/// One island, end to end. Returns an outcome rather than throwing, so the
/// caller's exit-code mapping stays exhaustive over the sealed type.
Future<CheckResult> check(String repoRoot, Island island) async {
  CheckResult cannot(String reason) =>
      (outcome: CannotRun(island.name, reason), values: _noValues);

  final repoRead = await decryptRepoEnv(repoRoot, island.name);
  if (repoRead case ReadFailed(:final reason)) return cannot(reason);
  final boxRead = await readLiveEnv(island);
  if (boxRead case ReadFailed(:final reason)) return cannot(reason);

  final Map<String, String> repo;
  final Map<String, String> box;
  try {
    repo = parseDotenv(
      (repoRead as ReadOk).text,
      'repo copy of ${island.name}',
    );
    box = parseDotenv(
      (boxRead as ReadOk).text,
      'live .env on ${island.sshHost}',
    );
  } on DotenvParseException catch (e) {
    return cannot('$e');
  }

  final wrong = wrongBox(island.name, island.sshHost, repo, box);
  if (wrong != null) return cannot(wrong);

  final d = diff(island.name, repo, box);
  if (d == null) {
    return (outcome: Match(island.name, repo.length), values: _noValues);
  }
  return (
    outcome: d,
    values: {for (final k in d.valuesDiffer) k: (repo[k]!, box[k]!)},
  );
}

void _report(CheckResult result, bool showValues) {
  switch (result.outcome) {
    case Match(:final island, :final keyCount):
      stdout.writeln('\n=== $island (${islands[island]!.sshHost}) ===');
      stdout.writeln(
        "  MATCH — $keyCount keys, the box's .env agrees with the repo's "
        'encrypted copy.',
      );

    case Differs(
      :final island,
      :final onlyInRepo,
      :final onlyOnBox,
      :final valuesDiffer,
    ):
      stdout.writeln('\n=== $island (${islands[island]!.sshHost}) ===');
      stdout.writeln('  DIFFERS');
      if (onlyInRepo.isNotEmpty) {
        stdout.writeln(
          '\n  in the REPO but NOT on the box (${onlyInRepo.length}):',
        );
        for (final k in onlyInRepo) {
          stdout.writeln('    - $k');
        }
      }
      if (onlyOnBox.isNotEmpty) {
        stdout.writeln(
          '\n  on the BOX but NOT in the repo (${onlyOnBox.length}):',
        );
        for (final k in onlyOnBox) {
          stdout.writeln('    + $k');
        }
      }
      if (valuesDiffer.isNotEmpty) {
        stdout.writeln(
          '\n  present in both, VALUES DIFFER (${valuesDiffer.length}):',
        );
        for (final k in valuesDiffer) {
          if (showValues) {
            final (r, b) = result.values[k]!;
            stdout.writeln('    ~ $k\n        repo: $r\n        box:  $b');
          } else {
            stdout.writeln('    ~ $k');
          }
        }
        if (!showValues) {
          stdout.writeln(
            '\n  (key names only — pass --show-values to print the values)',
          );
        }
      }

    case CannotRun(:final island, :final reason):
      stderr.writeln('\n=== $island ===\n  COULD NOT RUN — $reason');
  }
}

/// The repo root, derived from this script's own location rather than from the
/// caller's cwd — so the tool works from anywhere, which is how an operator will
/// actually invoke it.
String _repoRoot() {
  var dir = File.fromUri(Platform.script).absolute.parent; // bin/
  return dir.parent.parent.parent.path; // dart/ -> deploy/ -> repo root
}

const _usage = '''
usage:
  dart run deploy/dart/bin/check_env_drift.dart <island> [<island> ...] [--show-values]
  dart run deploy/dart/bin/check_env_drift.dart --all

islands: enspyr, imagineering
exit:    0 match, 1 differs, 2 could not run (fails closed)''';
